core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -3490,6 +3491,12 @@ private static AtomicLong getAtomicLong(String key) {
     return atomic;
   }
 
+  /** Support the ARRAY_DISTINCT function. */
+  public static List distinct(List list) {
+    Set result = new LinkedHashSet<>(list);
```

Comment:
I'm not super familiar with SQL distinct semantics, but this might be not enough:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala#L27

Reply:
this is because the binary compare https://github.com/apache/spark/blob/816ebaca4a81e0a4369b1ffff43a76e23f5e4271/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L4064, but you can refer flink implement i supports.

https://github.com/apache/flink/commit/b9f6a90a9f942e6a69bb31bbe8baf8ab463156d5